dex/networks/eth/abi.go

```diff
@@ -0,0 +1,166 @@
+	abidata, err := parseSelector(selector)
+	if err != nil {
+		return nil, err
+	}
+	// Parse the call data according to the requested selector
+	return parseCallData(calldata, string(abidata))
+}
+
+// selectorRegexp is used to validate that a 4byte database selector corresponds
+// to a valid ABI function declaration.
+//
+// Note, although uppercase letters are not part of the ABI spec, this regexp
+// still accepts it as the general format is valid. It will be rejected later
+// by the type checker.
+var selectorRegexp = regexp.MustCompile(`^([^\)]+)\(([A-Za-z0-9,\[\]]*)\)`)
+
+// parseSelector converts a method selector into an ABI JSON spec. The returned
+// data is a valid JSON string which can be consumed by the standard abi package.
+func parseSelector(unescapedSelector string) ([]byte, error) {
+	// Define a tiny fake ABI struct for JSON marshalling
+	type fakeArg struct {
+		Type string `json:"type"`
+	}
+	type fakeABI struct {
+		Name   string    `json:"name"`
+		Type   string    `json:"type"`
+		Inputs []fakeArg `json:"inputs"`
+	}
+	// Validate the unescapedSelector and extract it's components
+	groups := selectorRegexp.FindStringSubmatch(unescapedSelector)
+	if len(groups) != 3 {
+		return nil, fmt.Errorf("invalid selector %q (%v matches)", unescapedSelector, len(groups))
+	}
+	name := groups[1]
+	args := groups[2]
+
+	// Reassemble the fake ABI and constuct the JSON
+	arguments := make([]fakeArg, 0)
+	if len(args) > 0 {
+		for _, arg := range strings.Split(args, ",") {
+			arguments = append(arguments, fakeArg{arg})
+		}
+	}
+	return json.Marshal([]fakeABI{{name, "function", arguments}})
+}
+
+// parseCallData matches the provided call data against the ABI definition and
+// returns a struct containing the actual go-typed values.
+func parseCallData(calldata []byte, unescapedAbidata string) (*decodedCallData, error) {
+	// Validate the call data that it has the 4byte prefix and the rest divisible by 32 bytes
+	if len(calldata) < 4 {
+		return nil, fmt.Errorf("invalid call data, incomplete method signature (%d bytes < 4)", len(calldata))
+	}
+	sigdata := calldata[:4]
+
+	argdata := calldata[4:]
+	if len(argdata)%32 != 0 {
+		return nil, fmt.Errorf("invalid call data; length should be a multiple of 32 bytes (was %d)", len(argdata))
+	}
+	// Validate the called method and upack the call data accordingly
+	abispec, err := abi.JSON(strings.NewReader(unescapedAbidata))
+	if err != nil {
+		return nil, fmt.Errorf("invalid method signature (%q): %v", unescapedAbidata, err)
+	}
+	method, err := abispec.MethodById(sigdata)
+	if err != nil {
+		return nil, err
+	}
+	values, err := method.Inputs.UnpackValues(argdata)
+	if err != nil {
+		return nil, fmt.Errorf("signature %q matches, but arguments mismatch: %v", method.String(), err)
+	}
+	// Everything valid, assemble the call infos for the signer
+	decoded := decodedCallData{signature: method.Sig, name: method.RawName}
+	for i := 0; i < len(method.Inputs); i++ {
```

Comment:
We should probably just leave the code verbatim so we can more easily apply fixes/updates from upstream, but it bugs me they don't just `range`... or allocate `decoded.inputs`.  Oh well